sdk/communication/Azure.Communication.Calling.Server/src/ContentDownloader.cs

```diff
@@ -0,0 +1,143 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Communication.Calling.Server.Models;
+using Azure.Core;
+using Azure.Core.Pipeline;
+
+namespace Azure.Communication.Calling.Server
+{
+    internal class ContentDownloader
+    {
+        private readonly ConversationClient _client;
+
+        internal ContentDownloader(ConversationClient client)
+        {
+            _client = client;
+        }
+
+        internal async Task<Response<Stream>> DownloadStreamingInternal(Uri endpoint, HttpRange range, bool async, CancellationToken cancellationToken)
+        {
+            using DiagnosticScope scope = _client._clientDiagnostics.CreateScope($"{nameof(ConversationClient)}.{nameof(DownloadStreamingInternal)}");
+            scope.Start();
+
+            try
+            {
+                Argument.AssertNotNull(endpoint, nameof(endpoint));
+            }
+            catch (Exception ex)
+            {
+                scope.Failed(ex);
+                throw;
+            }
+
+            Response<Stream> response = await StartDownloadAsync(endpoint, range, async: async, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            Stream stream = RetriableStream.Create(
+                response.Value,
+                startOffset => StartDownloadAsync(endpoint, range, startOffset, async, cancellationToken).EnsureCompleted().Value,
```

Comment:
if `async` is `true`, then this code is not threadsafe. `EnsureCompleted` could cause a deadlock. `EnsureCompleted` should only be called if you know the task is completed.

Reply:
I was using the Storage's [BlobBaseClient](https://github.com/Azure/azure-sdk-for-net/blob/6c222f666dd7e040db5abee148fbab770c2011d6/sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs#L1047-L1058) as a model.
I thought that `EnsureCompleted` would wait until the task is completed.
What is your recommendation here? to include an `await` to force it to complete?

Reply:
`EnsureCompleted` waits, but `Task.Wait` could cause deadlock. So it should not be used for sync over async. Instead you need to have a sync implementation.

Reply:
I split the method into two different ones, removing `EnsureCompleted` from the picture.